streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -1872,37 +1864,35 @@ private boolean hasSubscriptionUpdates() {
     }
 
     synchronized void addSubscribedTopicsFromAssignment(final List<TopicPartition> partitions, final String logPrefix) {
-        if (sourceTopicPattern() != null) {
+        if (usesPatternSubscription()) {
             final Set<String> assignedTopics = new HashSet<>();
             for (final TopicPartition topicPartition : partitions) {
                 assignedTopics.add(topicPartition.topic());
             }
-
-            final Collection<String> existingTopics = subscriptionUpdates();
-            if (!existingTopics.containsAll(assignedTopics)) {
-                assignedTopics.addAll(existingTopics);
-                updateSubscribedTopics(assignedTopics, logPrefix);
-            }
+            updateSubscribedTopics(assignedTopics, logPrefix);
         }
     }
 
     synchronized void addSubscribedTopicsFromMetadata(final Set<String> topics, final String logPrefix) {
-        if (sourceTopicPattern() != null) {
-            final Collection<String> existingTopics = subscriptionUpdates();
-            if (!existingTopics.equals(topics)) {
-                topics.addAll(existingTopics);
-                updateSubscribedTopics(topics, logPrefix);
-            }
+        if (usesPatternSubscription()) {
+            updateSubscribedTopics(topics, logPrefix);
         }
     }
 
     private void updateSubscribedTopics(final Set<String> topics, final String logPrefix) {
-        log.debug("{}found {} topics possibly matching subscription", logPrefix, topics.size());
-        subscriptionUpdates.clear();
-        subscriptionUpdates.addAll(topics);
+        final Collection<String> existingTopics = subscriptionUpdates();
+
+        if  (usesPatternSubscription() && !existingTopics.equals(topics)) {
```

Comment:
Seems the callers are conditioned on `usesPatternSubscription()` already?

Reply:
Good catch